truffleruby/src/main/java/org/truffleruby/core/basicobject/BasicObjectNodes.java

```diff
@@ -126,9 +128,9 @@ public boolean equal(DynamicObject a, DynamicObject b) {
             return a == b;
         }
 
-        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "notSameClass(a, b)", "isNotIntLong(a) || isNotIntLong(b)" })
-        public boolean equalIncompatiblePrimitiveTypes(Object a, Object b) {
-            return false;
+        @Specialization(guards = { "isNotDynamicObject(a)", "isNotDynamicObject(b)", "isNotIntLong(a) || isNotIntLong(b)" })
+        public boolean equalrimitiveTypes(Object a, Object b) {
```

Comment:
This is introducing a bug. `equal?` should still be identity comparison for Java objects.